cmd/zoekt-dynamic-indexserver/main.go

```diff
@@ -0,0 +1,212 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This program manages a zoekt dynamic indexing deployment:
+// * listens to indexing commands
+// * reindexes specified repositories
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"time"
+)
+
+func loggedRun(cmd *exec.Cmd) (out, err []byte) {
+	outBuf := &bytes.Buffer{}
+	errBuf := &bytes.Buffer{}
+	cmd.Stdout = outBuf
+	cmd.Stderr = errBuf
+
+	log.Printf("run %v", cmd.Args)
+	if err := cmd.Run(); err != nil {
+		log.Printf("command %s failed: %v\nOUT: %s\nERR: %s",
+			cmd.Args, err, outBuf.String(), errBuf.String())
+	}
+
+	return outBuf.Bytes(), errBuf.Bytes()
+}
+
+type Options struct {
+	indexTimeout time.Duration
+	dataDir      string
+	indexDir     string
+	repoDir      string
+	listen       string
+}
+
+func (o *Options) createMissingDirectories() {
+	for _, s := range []string{o.dataDir, o.indexDir, o.repoDir} {
+		if _, err := os.Stat(s); err == nil {
+			continue
+		}
+
+		if err := os.MkdirAll(s, 0o755); err != nil {
+			log.Fatalf("MkdirAll %s: %v", s, err)
+		}
+	}
+}
+
+type indexRequest struct {
+	CloneURL string // TODO: Decide if tokens can be in the URL or if we should pass separately
+	RepoID   uint32
+}
+
+// This function is declared as var so that we can stub it in test
+var executeCmd = func(ctx context.Context, name string, arg ...string) {
+	cmd := exec.CommandContext(ctx, name, arg...)
+	cmd.Stdin = &bytes.Buffer{}
+	loggedRun(cmd)
+}
+
+func indexRepository(opts Options, req indexRequest, w http.ResponseWriter) {
+	ctx, cancel := context.WithTimeout(context.Background(), opts.indexTimeout)
+	defer cancel()
+
+	args := []string{}
+	args = append(args, "-dest", opts.repoDir)
+	args = append(args, "-name", strconv.FormatUint(uint64(req.RepoID), 10))
+	args = append(args, "-repoid", strconv.FormatUint(uint64(req.RepoID), 10))
+	args = append(args, req.CloneURL)
+	executeCmd(ctx, "zoekt-git-clone", args...)
+
+	gitRepoPath, err := filepath.Abs(filepath.Join(opts.repoDir, fmt.Sprintf("%d.git", req.RepoID)))
+	if err != nil {
+		log.Printf("error loading git repo path: %v", err)
+		http.Error(w, "JSON parser error", http.StatusBadRequest)
+		return
+	}
+
+	args = []string{
+		"-C",
+		gitRepoPath,
+		"fetch",
+	}
+	executeCmd(ctx, "git", args...)
+
+	args = []string{
+		"-index", opts.indexDir,
+		gitRepoPath,
+	}
+	executeCmd(ctx, "zoekt-git-index", args...)
+}
```

Comment:
you are not writing to w on success

Reply:
`serveIndex` now renders json on success https://github.com/sourcegraph/zoekt/pull/496/files#diff-d23d6eeb3bb1fdaf40622f2869e605ad254e69bac940b1e4a5b84fc590b3ada7R154